diff --git a/drivers/sensor/st/stm32_temp/stm32_temp.c b/drivers/sensor/st/stm32_temp/stm32_temp.c
index 0050985c2df..55904eccb61 100644
--- a/drivers/sensor/st/stm32_temp/stm32_temp.c
+++ b/drivers/sensor/st/stm32_temp/stm32_temp.c
@@ -138,6 +138,7 @@ static float convert_adc_sample_to_temperature(const struct device *dev)
 	/* Perform multiplication first for higher accuracy */
 	const int vsense = ((int)data->raw * vdda_mv) / 4095;
 
+	LOG_DBG("vdda_mv:%d, raw:%d, v25:%d", vdda_mv, data->raw, cfg->v25));
 	if (cfg->is_ntc) {
 		temperature = (float)(cfg->v25 - vsense);
 	} else {
@@ -173,6 +174,7 @@ static float convert_adc_sample_to_temperature(const struct device *dev)
 	} else {
 		dividend = (sense_data - (cd->ts_cal1 >> cfg->calib_data_shift));
 	}
+	LOG_DBG("vdda_mv:%d, sense_data:%d ave_slope_code:%d", vdda_mv, (int)sense_data, (int)ave_slope_code);
 
 	temperature = (dividend / avg_slope_code) + cfg->ts_cal1_temp;
 #else /* HAS_DUAL_CALIBRATION */
@@ -182,10 +184,19 @@ static float convert_adc_sample_to_temperature(const struct device *dev)
 	 *
 	 *                 (TS_CAL2_TEMP - TS_CAL1_TEMP)
 	 *  where Slope =  -----------------------------
-	 *                      (TS_CAL2 - TS_CAL1)
+	 *  					(TS_CAL2 - TS_CAL1)
+	 *
+	 * convert_adc_sample_to_temperature: vdda_mv:3300,
+	 * sense_data:1763 slope:0 ts_cal1_temp:30, ts_cal2_temp:110
+	 * convert_adc_sample_to_temperature: calib1:1310, calib0:1738,
+	 * calib_data_shift:0
 	 */
 	const float slope = ((float)(cfg->ts_cal2_temp - cfg->ts_cal1_temp))
-					/ ((cd->ts_cal2 - cd->ts_cal1) >> cfg->calib_data_shift);
+					/ (((int)cd->ts_cal2 - (int)cd->ts_cal1) >> cfg->calib_data_shift);
+	// LOG_DBG("vdda_mv:%d, sense_data:%d slope:%d ts_cal1_temp:%d, ts_cal2_temp:%d",
+	// 		vdda_mv, (int)sense_data, (int)slope, cfg->ts_cal1_temp, cfg->ts_cal2_temp);
+	// LOG_DBG("calib1:%d, calib0:%d, calib_data_shift:%d",
+	// 		calib[1], calib[0], cfg->calib_data_shift);
 
 	temperature = (slope * (sense_data - (cd->ts_cal1 >> cfg->calib_data_shift)))
 			+ cfg->ts_cal1_temp;
