From 97fc3b55d1aa165f8d8c857b87b80a640c5b46b2 Mon Sep 17 00:00:00 2001
From: Pete Skeggs <plskeggs@noeticdesign.com>
Date: Sun, 14 Jun 2026 18:23:07 -0700
Subject: [PATCH] Fix bug in patch

---
 stack/CO_PDO.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/stack/CO_PDO.c b/stack/CO_PDO.c
index 1696d39..7c29e55 100644
--- a/stack/CO_PDO.c
+++ b/stack/CO_PDO.c
@@ -149,7 +149,15 @@ static void CO_TPDOconfigCom(CO_TPDO_t* TPDO, uint32_t COB_IDUsedByTPDO, uint8_t
     /* is TPDO used? */
     if((COB_IDUsedByTPDO & 0xBFFFF800L) == 0 && TPDO->dataLength && ID){
         /* is used default COB-ID? */
-        if(ID == TPDO->defaultCOB_ID) ID += TPDO->nodeId;
+        if(ID == TPDO->defaultCOB_ID) {
+            ID += TPDO->nodeId;
+        } else {
+            int i = TPDO->SDO->entryNo;
+
+            if((ID & 0x7FF) == (0x180U + (0x100U * (i % 4) + (i / 4)))) {
+                ID += TPDO->nodeId;
+            }
+        }
         TPDO->valid = true;
     }
     else{
-- 
2.43.0

